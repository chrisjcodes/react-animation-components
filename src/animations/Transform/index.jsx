import React from 'react';
import { node, object, string } from 'prop-types';
import { Transition } from 'react-transition-group';

import {
    defaultAnimationProps,
    getInlineStyles,
    getTimeoutValue,
} from 'utilities';

const Transform = ({ children, enterTransform, exitTransform, ...props }) => {
    const pos = {
        entering: exitTransform,
        entered: enterTransform,
        exiting: exitTransform,
        exited: exitTransform,
    };

    return (
        <Transition timeout={getTimeoutValue(props)} {...props}>
            {status => (
                <div
                    className={props.className}
                    style={{
                        ...getInlineStyles(props),
                        transform: pos[status],
                        transitionProperty: 'transform',
                        willChange: 'transform',
                    }}
                >
                    {children}
                </div>
            )}
        </Transition>
    );
};

Transform.propTypes = {
    appear: bool,
    children: node.isRequired,
    className: string,
    delay: number,
    duration: number,
    enterTransform: string,
    exitTransform: string,
    style: object,
    timingFn: string,
};

Transform.defaultProps = {
    ...defaultAnimationProps,
    enterTransform: 'none',
    exitTransform: 'none',
};

export default Transform;
