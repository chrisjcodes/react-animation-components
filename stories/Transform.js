import React from 'react';

import { storiesOf } from '@storybook/react';
import * as knobs from '@storybook/addon-knobs';

import { createCommonKnobs } from '../src/utilities';

import { Transform } from '../src/index';

storiesOf('Animations/Transform', module)
    .addDecorator(knobs.withKnobs)
    .add('default', () => {
        const commonKnobs = createCommonKnobs(knobs);
        const enterTransform = knobs.text('enterTransform', 'translateY(50vh)');
        const exitTransform = knobs.text('exitTransform', 'none');

        return (
            <Transform
                in={commonKnobs.inProp}
                delay={commonKnobs.delay}
                duration={commonKnobs.duration}
                timingFn={commonKnobs.timingFn}
                enterTransform={enterTransform}
                exitTransform={exitTransform}
            >
                <h1>Example</h1>
            </Transform>
        );
    });
