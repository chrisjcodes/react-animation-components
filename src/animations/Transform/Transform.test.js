import React from 'react';
import Transform from './index.jsx';
import renderer from 'react-test-renderer';

describe('Transform', () => {
    test('renders', () => {
        const component = renderer.create(<Transform />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('can accept className', () => {
        const component = renderer.create(<Transform className="test" />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('can accept custom styles', () => {
        const component = renderer.create(
            <Transform style={{ background: 'black' }} />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('sets transitionDelay with delay prop', () => {
        const component = renderer.create(<Transform delay={1000} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('sets transitionDuration with duration prop', () => {
        const component = renderer.create(<Transform duration={1000} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('sets transitionTimingFunction with timingFn prop', () => {
        const component = renderer.create(<Transform timingFn="ease-in-out" />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('sets enterTransform', () => {
        const component = renderer.create(
            <Transform enterTransform="translateX(10em)" />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('sets exitTransform', () => {
        const component = renderer.create(
            <Transform exitTransform="translateX(10em)" />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
