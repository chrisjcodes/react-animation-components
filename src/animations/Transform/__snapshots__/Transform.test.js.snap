// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform can accept className 1`] = `
<div
  className="test"
  style={
    Object {
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform can accept custom styles 1`] = `
<div
  className={undefined}
  style={
    Object {
      "background": "black",
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform renders 1`] = `
<div
  className={undefined}
  style={
    Object {
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform sets finish 1`] = `
<div
  className={undefined}
  style={
    Object {
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform sets start 1`] = `
<div
  className={undefined}
  style={
    Object {
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform sets transitionDelay with delay prop 1`] = `
<div
  className={undefined}
  style={
    Object {
      "transform": "none",
      "transitionDelay": "1000ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform sets transitionDuration with duration prop 1`] = `
<div
  className={undefined}
  style={
    Object {
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "1000ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease",
      "willChange": "transform",
    }
  }
/>
`;

exports[`Transform sets transitionTimingFunction with timingFn prop 1`] = `
<div
  className={undefined}
  style={
    Object {
      "transform": "none",
      "transitionDelay": "0ms",
      "transitionDuration": "500ms",
      "transitionProperty": "transform",
      "transitionTimingFunction": "ease-in-out",
      "willChange": "transform",
    }
  }
/>
`;
